suzieq/shared/sqPlugin.py

```diff
@@ -0,0 +1,66 @@
+import abc
+from importlib import import_module
+from importlib.util import find_spec
+from inspect import getmembers, getmro, isclass
+from pkgutil import iter_modules
+from typing import Dict, Type
+
+
+class SqPlugin:
+    """SqPlugin is the base common class inherited by all the
+    Suzieq plugins
+    """
+
+    @abc.abstractmethod
+    def __init__(self) -> None:
+        pass
+
+    @classmethod
+    def discover_plugins(cls,
+                         search_pkg: str,
+                         transitive=False) -> Dict[str, Type]:
+        """Discover all the plugins in the search_pkg package, inheriting
+        the current base class.
+
+        Args:
+            search_pkg (str): The package where to look for the plugins.
+                Note that if the specified package contains other packages,
+                the function will scan only their content, without accessing
+                to other nested packages and ignoring all the files inside
+                search_pkg.
+            transitive (bool, optional): If False returns a plugin only if
+                it is a direct descendant of the base class.
+                Defaults to False.
+
+        Returns:
+            Dict[str, Type]: a dictionary containing all
+                the descendants of the base class in the search_pkg,
+                with format {'module name': PluginClass}
+        """
+        classes = {}
+
+        # Collect the list of packages where to search
+        search_path = search_pkg.replace('.', '/')
+        packages = [f'{search_pkg}.{m.name}'
+                    for m in iter_modules([search_path])
+                    if m.ispkg]
+        if not packages:
+            packages.append(search_pkg)
+
+        for pkg in packages:
+            pspec = find_spec(pkg)
+            if pspec and pspec.loader:
+                mfound = [x.split('.')[0] for x in pspec.loader.contents()
+                          if not x.startswith('_')]
+                for x in mfound:
+                    use_name = x.split('.')[0]
+                    mname = f'{pkg}.{use_name}'
+                    m = import_module(mname)
+                    for mbr in getmembers(m, isclass):
+                        if (mbr[1].__module__ == mname
+                           and (transitive or getmro(mbr[1])[1] == cls)
+                           and issubclass(mbr[1], cls)
+                           and mbr[1] != cls):
+                            classes[use_name] = mbr[1]
+
+        return classes
```

Comment:
Do we need an abstract property or method called "plugin_type" that identifies the plugin type?

Reply:
As we discover by the class I was thinking about a solution using the Class type to identify a plugin type. I'm sure having a property allows to easily know the type of the plugin once you have one of it, but at point shouldn't we have a function allowing to discover the plugin by the value of the property "type"?

Reply:
Thats interesting. Can you propose some code for this?

Reply:
The code does not change so much, I think it should be enough declaring an class attribute like:
```
class SqPlugin:
         plugin_type = "BasePlugin"
         ...
```
Overriding this attribute every time we inherit SqPlugin from a plugin base class (i.e. SqObject).
From the point of view of the the `get_plugin()` method, we don't look anymore to the class but on the plugin_type string. I.e.
```
...
for mbr in getmembers(mod, isclass):
    if (mbr[1].__module__ == mname
        mbr[1].plugin_type == cls.plugin_type
        and mbr[1] != cls):
        classes[use_name] = mbr[1]
...
``` 
At this point, this solution allows to easily retrieve the kind of plugin by getting the plugin_type class attribute, but I think we might create some problems than the ones we solve:

- We need to manually set the `plugin_type` on the plugin base classes (i.e. in the SqObjects class), but it might be prone to errors (plugin_type not overwritten, typos, "service" instead of "Service"...)
- Classes having different parents but with the same `plugin_type` attribute are considered the same type of plugin, is it logically correct?